✨ Student Review Component

b-l-i-n-d opened the pull request:
In this PR:
- Make a dynamic review component with API integrations.
- ✨ New `StarRatingInput`, `StartRating` components
- `InputField` component bugfix
- `Button` component icon-only variant

b-l-i-n-d committed:
feat: Make profile header sticky, standardize dashboard navigation permalinks, and refactor the given reviews template.

b-l-i-n-d committed:
feat: Enhance the Button component with an icon-only option.

b-l-i-n-d committed:
feat: Implement dynamic review card rendering with structured data, add review action transition, and remove debug output.

b-l-i-n-d committed:
Merge branch '4.0.0-dev' into v4-reviews

b-l-i-n-d committed:
fix: Error and clear button not rendering properly

b-l-i-n-d committed:
feat: new Icons added

b-l-i-n-d committed:
feat: Interactive star rating component

b-l-i-n-d committed:
feat: allow users to manage their given reviews in the dashboard

b-l-i-n-d committed:
feat: Add frontend dashboard page for managing user reviews with edit/delete functionality, new API endpoints, and associated UI components.

b-l-i-n-d committed:
Merge branch '4.0.0-dev' into v4-reviews

b-l-i-n-d committed:
feat: Add pagination and empty state to given reviews dashboard and standardize review form field names.

b-l-i-n-d committed:
chore: Update imports

b-l-i-n-d committed:
chore: Remove unused codes

b-l-i-n-d committed:
refactor: Remove unused codes

b-l-i-n-d committed:
refactor: Remove unnecessary exports

b-l-i-n-d committed:
Merge branch '4.0.0-dev' into v4-reviews

b-l-i-n-d committed:
refactor: Use tutor-transition mixing

b-l-i-n-d committed:
Merge branch '4.0.0-dev' into v4-reviews

b-l-i-n-d committed:
refactor: reimplement review deletion modal using ConfirmationModal component and update associated frontend logic

b-l-i-n-d committed:
feat: Add loading state to delete review button and reload page after successful review deletion.

sazedul-haque commented on components/Button.php:
```diff
@@ -216,6 +225,20 @@ public function icon( string $icon, string $position = 'left', int $width = 16,
 		return $this;
 	}
 
+	/**
```
We can determine if icon only or not programmatically. No need to add prop support.

b-l-i-n-d replied:
`Prop` support is needed, as label will be used as `aria-label` for accessibility.

sazedul-haque commented on components/Button.php:
```diff
@@ -304,9 +334,13 @@ public function get(): string {
-		$content = self::POSITION_RIGHT === ( $this->icon_position ? $this->icon_position : self::POSITION_LEFT )
-			? sprintf( '%1$s%2$s', esc_html( $this->label ), $icon_html )
-			: sprintf( '%1$s%2$s', $icon_html, esc_html( $this->label ) );
+		if ( $this->icon_only ) {
```
This is not necessary I guess.

sazedul-haque commented on templates/dashboard/reviews/given-reviews.php:
```diff
@@ -1,162 +1,107 @@
+				->message( __( 'Are you sure you want to delete this review? Please confirm your choice.', 'tutor' ) )
+				->confirm_handler( 'handleDeleteReview(payload?.id)' )
+				->mutation_state( 'deleteReviewMutation' )
+				->confirm_button(
```
No need to provide confirm_button and cancel_button. the default should work.

sazedul-haque commented on assets/src/js/frontend/dashboard/pages/reviews.ts:
```diff
@@ -0,0 +1,148 @@
+      return wpAjaxInstance.post(endpoints.PLACE_RATING, payload).then((res) => res.data);
+    },
+
+    convertFormDataToPayload(data: ReviewFormProps): ReviewPayload {
```
Try to skip this conversion.

b-l-i-n-d replied:
This is needed as the payload does not match the input names.

b-l-i-n-d commented on assets/src/scss/frontend/dashboard/_reviews.scss:
```diff
@@ -12,9 +12,7 @@
-    transition-property: background-color, border-color, opacity;
-    transition-duration: 0.2s;
-    transition-timing-function: ease-in-out;
+    @include tutor-transition((background-color, border-color, opacity));
```
This is the params of the `mixing`, without `brackets` it will be considered as three different params.

b-l-i-n-d commented on templates/dashboard/reviews/review-card.php:
```diff
@@ -0,0 +1,200 @@
+			@submit.prevent="handleSubmit(
+				(data) => handleReviewSubmit(data),
+				(errors) => { 
+					console.log('Form errors:', errors); 
```
This is not intentional. Removing this since we are handling errors inside `query`

b-l-i-n-d committed:
feat: swap confirm and cancel button styles and order in ConfirmationModal

b-l-i-n-d committed:
feat: introduce StarRating component and integrate it into the review card template.

b-l-i-n-d committed:
refactor: rename `showAverage` and `iconSize` methods to `show_average` and `icon_size` for consistency.

b-l-i-n-d committed:
chore: remove reviews demo component and its associated template file.

b-l-i-n-d committed:
style: update star rating average and count text styling classes

b-l-i-n-d commented on templates/demo-components/dashboard/components/star-rating.php:
```diff
@@ -12,27 +12,32 @@
 if ( ! defined( 'ABSPATH' ) ) {
 	exit;
 }
+use TUTOR\Icon;
```
Removed demo component and created new `StarRating` component

b-l-i-n-d committed:
refactor: replace `confirm_button` and `cancel_button` methods with `confirm_text` and `cancel_text` for modal actions.

b-l-i-n-d committed:
style: Update confirmation button class from primary to destructive.

b-l-i-n-d committed:
refactor: Align review data structure with WordPress comment fields and remove redundant data conversion.

b-l-i-n-d committed:
fix: Use `comment_ID` and `comment_content` for review data instead of `id` and `review_content`.

b-l-i-n-d committed:
Merge branch '4.0.0-dev' into v4-reviews

sazedul-haque commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
+			</table>
+		</div>
+	<?php else : ?>
+		<?php tutor_utils()->tutor_empty_state( tutor_utils()->not_found_text() ); ?>
```
Please use empty state component

b-l-i-n-d replied:
Same here

sazedul-haque commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
+	<?php endif; ?>
+
+	<?php if ( $reviews->count ) : ?>
+		<div class="tutor-table-responsive">
```
Where is this design?

harunollyo commented on components/InputField.php:
```diff
@@ -515,20 +505,6 @@ public function collapsable( $close_on_select = true ): self {
-	 *
-	 * @return $this
-	 */
-	public function error( $error ) {
```
Why this method removed?

harunollyo commented on components/StarRating.php:
```diff
@@ -0,0 +1,164 @@
+						</span>
+					<?php endif; ?>
+
+					<?php if ( null !== $this->count ) : ?>
```
We write simply
`if ( $this->count )`

harunollyo commented on components/StarRating.php:
```diff
@@ -0,0 +1,164 @@
+				<?php endfor; ?>
+			</div>
+
+			<?php if ( $this->show_average || null !== $this->count ) : ?>
```
```php
if ( $this->show_average ||  $this->count )
```

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,176 @@
+	 *
+	 * @var string
+	 */
+	protected $field_name = '';
```
The default name should be `rating`

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,176 @@
+	 *
+	 * @var int
+	 */
+	protected $icon_size = 20;
```
We can use `Size::SIZE_20`

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,176 @@
+		<div 
+			class="tutor-flex tutor-gap-4 tutor-justify-between tutor-items-center"
+			x-data="tutorStarRatingInput({
+				initialRating: <?php echo esc_js( $current_rating ); ?>,
```
Use `esc_attr`

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,176 @@
+					<button 
+						type="button"
+						class="tutor-btn tutor-btn-link tutor-p-none tutor-min-h-0"
+						@click="setRating(<?php echo (int) $i; ?>, (rating) => setValue('<?php echo esc_attr( $this->field_name ); ?>', rating))"
```
`$i` is already int, no need to cast here

b-l-i-n-d replied:
While $i is technically an integer from the loop logic, PHPCS strict standards (like WordPress-Extra) often require all output to pass through an escaping function or be explicitly cast to prove safety.

harunollyo replied:
In this case we should use `esc_attr` instead of re-casting

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,176 @@
+						type="button"
+						class="tutor-btn tutor-btn-link tutor-p-none tutor-min-h-0"
+						@click="setRating(<?php echo (int) $i; ?>, (rating) => setValue('<?php echo esc_attr( $this->field_name ); ?>', rating))"
+						@mouseenter="hoverRating = <?php echo (int) $i; ?>"
```
same here

b-l-i-n-d replied:
While $i is technically an integer from the loop logic, PHPCS strict standards (like WordPress-Extra) often require all output to pass through an escaping function or be explicitly cast to prove safety.

harunollyo replied:
In this case we should use `esc_attr` instead of re-casting

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,176 @@
+						@click="setRating(<?php echo (int) $i; ?>, (rating) => setValue('<?php echo esc_attr( $this->field_name ); ?>', rating))"
+						@mouseenter="hoverRating = <?php echo (int) $i; ?>"
+					>
+						<template x-if="effectiveRating >= <?php echo (int) $i; ?>">
```
same here

b-l-i-n-d replied:
While $i is technically an integer from the loop logic, PHPCS strict standards (like WordPress-Extra) often require all output to pass through an escaping function or be explicitly cast to prove safety.

harunollyo replied:
In this case we should use esc_attr instead of re-casting

harunollyo commented on components/StarRating.php:
```diff
@@ -0,0 +1,164 @@
+		$star_half = tutor_utils()->get_svg_icon( Icon::STAR_HALF, $icon_size, $icon_size );
+		$star      = tutor_utils()->get_svg_icon( Icon::STAR_2, $icon_size, $icon_size );
+
+		ob_start();
```
Instead of output buffering set the prepared HTML to `$this->component_string`

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,176 @@
+		$star_half      = tutor_utils()->get_svg_icon( Icon::STAR_HALF, $this->icon_size, $this->icon_size );
+		$star           = tutor_utils()->get_svg_icon( Icon::STAR_2, $this->icon_size, $this->icon_size );
+
+		ob_start();
```
Instead of output buffering set the prepared HTML to `$this->component_string`

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
  * @author Themeum <support@themeum.com>
  * @link https://themeum.com
- * @since 4.0.0
+ * @since 1.4.3
```
Keep it 4.0.0

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
 
 defined( 'ABSPATH' ) || exit;
 
+tutor_load_template(
```
Remove this and add this into your template
`<?php require_once tutor_get_template( 'account-header' ); ?>`

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
+	return;
+}
+
+use TUTOR\Input;
```
Move import statement on top

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
+use TUTOR\Input;
+
+// Pagination Variable.
+$per_page     = tutor_utils()->get_option( 'pagination_per_page', 20 );
```
Use `$item_per_page`
`$per_page ` gives CS error `Overriding WordPress globals is prohibited`

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
+$current_page = max( 1, Input::get( 'current_page', 1, Input::TYPE_INT ) );
+$offset       = ( $current_page - 1 ) * $per_page;
+
+$reviews     = tutor_utils()->get_reviews_by_instructor( get_current_user_id(), $offset, $per_page );
```
Instead of multiple function call, make `$user_id = get_current_user_id()` and use it where required.

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
+			<ul class="tutor-nav">
+				<li class="tutor-nav-item">
+					<a class="tutor-nav-link is-active" href="<?php echo esc_url( tutor_utils()->get_tutor_dashboard_page_permalink( 'reviews' ) ); ?>"> 
+						<?php esc_html_e( 'Received', 'tutor' ); ?> (<?php echo esc_html( $reviews->count ); ?>)
```
Make it translatable `'Received (%d)`

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
+				<?php if ( $given_count ) : ?>
+					<li class="tutor-nav-item"> 
+						<a class="tutor-nav-link" href="<?php echo esc_url( tutor_utils()->get_tutor_dashboard_page_permalink( 'reviews/given-reviews' ) ); ?>"> 
+							<?php esc_html_e( 'Given', 'tutor' ); ?> (<?php echo esc_html( $given_count ); ?>)
```
Same here

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -1,13 +1,133 @@
+					<tr>
+						<td class="tutor-td-top">
+							<div class="tutor-d-flex tutor-align-center">
+								<?php echo wp_kses( tutor_utils()->get_tutor_avatar( $review->user_id ), tutor_utils()->allowed_avatar_tags() ); ?>
```
Why we are not using new Avatar component here?

harunollyo commented on templates/dashboard/reviews/given-reviews.php:
```diff
@@ -1,162 +1,77 @@
-$per_page     = tutor_utils()->get_option( 'pagination_per_page', 20 );
-$current_page = max( 1, Input::get( 'current_page', 0, Input::TYPE_INT ) );
-$offset       = ( $current_page - 1 ) * $per_page;
+$pagination_per_page = tutor_utils()->get_option( 'pagination_per_page', 20 );
```
Use consistent variable name `$item_per_page`

harunollyo commented on templates/dashboard/reviews/review-card.php:
```diff
@@ -0,0 +1,191 @@
+ * @link https://themeum.com
+ * @since 4.0.0
+ */
+
```
Prevent direct access

harunollyo commented on templates/demo-components/dashboard/components/profile-pages-header.php:
```diff
@@ -10,21 +10,23 @@
 
 use TUTOR\Icon;
```
Revert this file changes. This is not part of this PR

harunollyo commented on templates/demo-components/dashboard/components/review-card.php:
```diff
@@ -1,87 +0,0 @@
-<?php
```
Why this file has been removed?

harunollyo commented on templates/demo-components/playground.php:
```diff
@@ -43,7 +43,6 @@
 		'learning-area'      => 'Learning Area',
 		'user-profile'       => 'User Profile',
 		'certificates'       => 'Certificates',
-		'reviews'            => 'Reviews',
```
Do not remove

harunollyo commented on templates/demo-components/reviews.php:
```diff
@@ -1,27 +0,0 @@
-<?php
-/**
```
Do not remove demo component now

b-l-i-n-d committed:
refactor: Refactor star rating components and reviews template

Updated StarRating and StarRatingInput components for improved logic and code consistency, including default field name, icon size constant, and safer value checks. Refactored the dashboard reviews template to streamline variable naming, update template loading, and improve pagination logic.

b-l-i-n-d committed:
refactor: Store rendered component string in property

Both StarRating and StarRatingInput now assign the rendered output to a component_string property before returning it. This allows access to the rendered string elsewhere in the component lifecycle if needed.

b-l-i-n-d committed:
feat: Add error message support to InputField component

Introduces an $error property and an error() setter method to the InputField component, allowing error messages to be set and displayed. Updates usage example in the docblock to demonstrate the new error() method.

b-l-i-n-d committed:
refactor: Refactor account header to use Button component

Replaced anchor tags with the Tutor Button component for the back and close actions in the account header. This improves consistency and leverages the component-based approach for buttons.

b-l-i-n-d committed:
refactor: Refactor dashboard reviews to unify received and given views

Merged the logic for displaying received and given reviews into a single template (reviews.php) and removed the now-redundant given-reviews.php. Updated the reviews page to dynamically handle both instructor and user reviews, streamline pagination, and improve the empty state handling.

b-l-i-n-d committed:
feat: Add editable state to review card actions and form

Actions and edit form for reviews are now conditionally rendered based on a new 'is_editable' property. Also updated review content sanitization and changed form mode to 'onChange' for edit mode.

b-l-i-n-d committed:
refactor: Refactor reviews pagination and is_editable logic

Simplified the assignment of the is_editable property for reviews by replacing array_map with a foreach loop. Removed unused received_count variable and improved variable naming for clarity in pagination and review count handling.

b-l-i-n-d committed:
feat: Improve initials generation in Avatar component

Refines the initials() method to generate initials from the user's display name by extracting the first letter of each word, uppercasing them, and limiting the result to two characters. This ensures more accurate and consistent avatar initials.

b-l-i-n-d committed:
feat: Display student info in review cards

Added student avatar and display name to review cards for non-editable reviews. Updated SCSS for student info styling and set user_avatar in reviews data.

b-l-i-n-d committed:
chore: Updated comment for review card

b-l-i-n-d committed:
refactor: Add review card component and reviews page template

Introduced a new review card component for displaying individual course reviews and created a reviews page template that utilizes this component. Minor formatting update was also made to the profile pages header component.

b-l-i-n-d committed:
chore: Add missing newlines at end of template files

Appended a newline character at the end of review-card.php and reviews.php to follow POSIX standards and improve file formatting consistency.

b-l-i-n-d committed:
refactor: Simplify initials processing in Avatar component

Refactored the initials() method to directly use the sanitized and uppercased input, removing logic that extracted initials from a display name. This change assumes the input is already the desired initials.

b-l-i-n-d committed:
refactor: Replace anchor tags with buttons in profile header

Changed navigation elements from <a> tags to <button> elements in the profile pages header component for improved accessibility and semantic correctness.

b-l-i-n-d committed:
chore: Add newline at end of profile header file

Appends a newline character at the end of profile-pages-header.php to improve file formatting and comply with POSIX standards.

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,179 @@
+												<?php echo $star_fill; // phpcs:ignore ?>
+							</span>
+						</template>
+						<template x-if="effectiveRating > <?php echo (int) ( $i - 1 ); ?> && effectiveRating < <?php echo (int) $i; ?>">
```
Same

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,179 @@
+												<?php echo $star_half; // phpcs:ignore ?>
+							</span>
+						</template>
+						<template x-if="effectiveRating <= <?php echo (int) ( $i - 1 ); ?>">
```
same

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -9,5 +9,84 @@
+				<?php endforeach; ?>
+			</div>
+
+			<?php if ( $review_count > $item_per_page ) : ?>
```
No need this, Pagination class will handle this.

harunollyo replied:
Remove this

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -9,5 +9,84 @@
+			<?php if ( $review_count > $item_per_page ) : ?>
+				<div class="tutor-mt-6">
+					<?php
+						Pagination::make()
```
Use `attr( 'class', 'tutor-mt-6' )`

harunollyo replied:
Pass as attr

b-l-i-n-d committed:
refactor: Sanitize output and update pagination in reviews

Replaced (int) casts with esc_attr for better output sanitization in StarRatingInput. Refactored reviews pagination to always render and added a class attribute for consistent styling.

b-l-i-n-d committed:
Merge '4.0.0-dev' into 'v4-reviews'

b-l-i-n-d committed:
Merge branch '4.0.0-dev' of https://github.com/themeum/tutor into v4-reviews

b-l-i-n-d committed:
Merge branch '4.0.0-dev' into v4-reviews

shewa12 commented on components/StarRating.php:
```diff
@@ -0,0 +1,166 @@
+					<span class="tutor-icon-exception4 tutor-flex-center">
+						<?php
+						if ( (int) $rating >= $i ) {
+							echo $star_fill; // phpcs:ignore
```
Why did not use the escape functions?

b-l-i-n-d replied:
`tutor_utils()->get_svg_icon` returns escaped value.

b-l-i-n-d committed:
chore: add comments explaining phpcs ignore for SVG icon echoes in StarRating.

sazedul-haque commented:
Wait for approval from harun bhai.

harunollyo commented on components/InputField.php:
```diff
@@ -1601,25 +1559,34 @@ public function get(): string {
+				x-show="!errors?.%1$s?.message"
+			>%2$s</div>',
+			esc_attr( $this->name ),
+			$this->esc( $this->help_text )
```
Use consistent method. Either wp helper or base.

`esc_html`

harunollyo commented on components/StarRating.php:
````diff
@@ -0,0 +1,166 @@
+ * ```
+ * StarRating::make()
+ *     ->rating(4.5)
+ *     ->showAverage(true)
````
Use sanke_case

harunollyo commented on components/StarRatingInput.php:
````diff
@@ -0,0 +1,179 @@
+ * Example Usage:
+ * ```
+ * StarRatingInput::make()
+ *     ->fieldName('rating')
````
Use snake_case

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,179 @@
+	 *
+	 * @return self
+	 */
+	public function fieldName( string $name ): self {
```
Use snake_case

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,179 @@
+	 *
+	 * @return self
+	 */
+	public function currentRating( float $rating ): self {
```
Use snake_case

harunollyo commented on components/StarRatingInput.php:
```diff
@@ -0,0 +1,179 @@
+	 *
+	 * @return self
+	 */
+	public function onChange( string $callback ): self {
```
Use snake_case

harunollyo commented on templates/dashboard/account/reviews.php:
```diff
@@ -9,5 +9,81 @@
+use Tutor\Components\Constants\Size;
+use Tutor\Components\Constants\Variant;
+
 defined( 'ABSPATH' ) || exit;
```
Use this immediate after file comment

harunollyo commented on components/StarRating.php:
```diff
@@ -0,0 +1,166 @@
+
+use TUTOR\Icon;
+
+defined( 'ABSPATH' ) || exit;
```
Use this immediate after file comment. Update all place where it is not in correct place.

harunollyo commented on components/StarRating.php:
```diff
@@ -0,0 +1,166 @@
+
+	/**
+	 * Rating value
+	 *
```
Use @since tag to all props and methods

b-l-i-n-d committed:
Merge branch '4.0.0-dev' into v4-reviews

b-l-i-n-d committed:
refactor: Remove PHP_INT_MAX

b-l-i-n-d committed:
refactor: rename StarRating and StarRatingInput methods to snake_case and update help text escaping in InputField.

b-l-i-n-d committed:
docs: Add `@since 4.0.0` tags to properties and methods in StarRating and StarRatingInput components.

harunollyo commented on templates/account-header.php:
```diff
@@ -9,22 +9,40 @@
+use Tutor\Components\Constants\Size;
+use Tutor\Components\Constants\Variant;
 
 defined( 'ABSPATH' ) || exit;
```
Move this immediate after file comment
